### PR TITLE
HDDS-6455. Adding a Builder class to Freon ContentGenerator, so as to avoid Constructor Telescoping.

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

The ContentGenerator class of the freon suite has multiple Parametrised Constructors each having various different arguments that have to be initialised, this affects the readability of the code. One way to solve this is to incorporate a builder class for ContentGenerator class so that any new parameters added in the future can be seamlessly added without the need of creating a new constructor.

When a file was closed or flushed, the file contents were written from the DataNode into the operating system using the normal close() and write() system calls.

The data may not be immediately persisted to the underlying physical storage, but may still reside in-memory in the operating system's file cache. This creates a window of vulnerability where if multiple DataNode machines fail simultaneously (e.g. loss of power to a rack), then previously written data may be lost. To combat this problem, HDFS (as of Hadoop 2.0) has introduced new APIs to provide a way to guarantee that written data will be immediately persisted to the underlying physical storage. These APIs are described in the following table.

> **hflush()** : This Method present inside the FSDataOutputStream.java flushes all outstanding data (i.e. the current unfinished packet) from the client into the OS buffers on all DataNode replicas.
> 
> **hsync()** : This Method present inside the FSDataOutputStream.java flushes the data to the DataNodes, like hflush(), but should also force the data to underlying physical storage via fsync (or equivalent).

Hence I have added a method that would implement Hsync or Hflush after every write if the user wants !

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6455

## How was this patch tested?

Existing UTs for ContentGeneratorClass ran successfully

### CODE REVIEW

File: hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java [hunk 1]
```diff
@@ -48,7 +49,10 @@ public void writeWithSmallerBuffers() throws IOException {
 
   @Test
   public void writeWithByteLevelWrite() throws IOException {
-    ContentGenerator generator = new ContentGenerator(1024, 1024, 1);
+    ContentGenerator generator =
+        new ContentGenerator.Builder().keySize(1024).bufferSize(1024)
+            .copyBufferSize(
+                1).build();
```
Comment (kaijchen) [inline hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java:55]: Redundant newline here.

File: hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java [hunk 2]
```diff
@@ -57,7 +61,10 @@ public void writeWithByteLevelWrite() throws IOException {
 
   @Test
   public void writeWithSmallBuffer() throws IOException {
-    ContentGenerator generator = new ContentGenerator(1024, 1024, 10);
+    ContentGenerator generator =
+        new ContentGenerator.Builder().keySize(1024).bufferSize(1024)
+            .copyBufferSize(
+                10).build();
```
Comment (kaijchen) [inline hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java:67]: Redundant newline here.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java [hunk 3]
```diff
@@ -154,14 +155,14 @@ private void createDir(long counter) throws Exception {
...
-            timer.getCount();
+        "Count=" + totalDirsCnt.get() + ", Total Files Count=" +
+        timer.getCount();
     print(message);
   }
 
   private void createSubDirRecursively(String parent, int depthIndex,
                                        int spanIndex)
-          throws Exception {
+      throws Exception {
```
Comment (kaijchen) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java:165]: Can be merged with previous line. (removing the vertical-align)

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java [hunk 4]
```diff
@@ -48,38 +49,100 @@
...
+   */
+  public static class Builder {
+
+    private long keySize = 1024;
+    private int bufferSize = 1024;
+    private int copyBufferSize = 1024;
+    private boolean hSync = false;
+    private boolean hFlush = false;
+
+    public Builder keySize(long keysize) {
```
Comment (kaijchen) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:114]: Is it better to use `setKeySize()` as the setter name here?

Comment (ArafatKhan2198) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:114]: Yea that'll help !! thanks for the suggestion

File: hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java (hunk 2 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java:67]: Sorry about that, Ill make the necessary changes !!

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java (hunk 3 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java:165]: Sorry about that, Ill make the necessary changes !!

**Comment by ArafatKhan2198** [comment]
> Thanks for the patch @ArafatKhan2198, I have left some comments for you to check.

@kaijchen all changes done !!

File: hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java [hunk 5]
```diff
@@ -57,7 +61,9 @@ public void writeWithByteLevelWrite() throws IOException {
 
   @Test
   public void writeWithSmallBuffer() throws IOException {
-    ContentGenerator generator = new ContentGenerator(1024, 1024, 10);
+    ContentGenerator generator =
+        new ContentGenerator.Builder().seyKeySize(1024).setBufferSize(1024)
+            .setCopyBufferSize(10).build();
```
Comment (jojochuang) [inline hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java:66]: typically builder classes are initialized one setter at a line. For example

```suggestion
        new ContentGenerator.Builder()
            .seyKeySize(1024)
            .setBufferSize(1024)
            .setCopyBufferSize(10)
            .build()
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java [hunk 6]
```diff
@@ -39,7 +39,7 @@
     aliases = "dfs-tree-generator",
     description =
         "Create nested directories and create given number of files in each " +
-                "dir in any dfs compatible file system.",
+            "dir in any dfs compatible file system.",
```
Comment (jojochuang) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java:42]: next time, let's not change the style of the unrelated code.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java [hunk 7]
```diff
@@ -48,38 +49,100 @@ public class ContentGenerator {
...
+    for (long nrRemaining = keySize; nrRemaining > 0;
+         nrRemaining -= bufferSize) {
       int curSize = (int) Math.min(bufferSize, nrRemaining);
       if (copyBufferSize == 1) {
         for (int i = 0; i < curSize; i++) {
           outputStream.write(buffer[i]);
+          flushOrSync(outputStream);
         }
-      } else {
+      } else {  /**
```
Comment (jojochuang) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:82]:
```suggestion
      } else {
        /**
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java (hunk 6 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopDirTreeGenerator.java]: Sorry about that, I have reverted the change !!

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java (hunk 7 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java]: Sorry about that, I have made the change !!

File: hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java (hunk 5 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/freon/TestContentGenerator.java]: Thanks for pointing it out!!
I have taken care of it across all the implementations !!

**Review by jojochuang — APPROVED** [review]

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java [hunk 8]
```diff
@@ -48,38 +49,101 @@ public class ContentGenerator {
...
+
+    public Builder setHflushData(boolean hflush) {
+      this.hFlush = hflush;
+      return this;
+    }
+
+    //Return the final constructed builder object
+    public ContentGenerator build() {
+      ContentGenerator contentgenerator = new ContentGenerator(this);
+      return contentgenerator;
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:143]: Builder's `build` method should verify that the object being constructed is valid.  Here `hflush` and `hsync` are not applicable at the same time.  Please add a check for that.

Nit: no need for local variable, return directly.

```suggestion
      return new ContentGenerator(this);
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java [hunk 9]
```diff
@@ -89,7 +90,11 @@ public Void call() throws Exception {
     }
 
     contentGenerator =
-        new ContentGenerator(fileSize, bufferSize, copyBufferSize);
+        new Builder()
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java:93]: Nit: I think it's more readable if builder is qualified where it is being used (and also more consistent with usage in other generator classes).

```suggestion
        new ContentGenerator.Builder()
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java [hunk 10]
```diff
@@ -28,6 +28,7 @@
 import org.apache.hadoop.hdds.conf.OzoneConfiguration;
 
 import com.codahale.metrics.Timer;
+import org.apache.hadoop.ozone.freon.ContentGenerator.Builder;
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java:31]:
```suggestion
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java [hunk 11]
```diff
@@ -48,38 +49,125 @@ public class ContentGenerator {
...
+
+    /**
+     * Type of flags permitted
+     * 1. hSync
+     * 2. hFlush
+     * 3. None.
+     */
+    public Builder setFlushMode(String flushmode) {
+      this.flushMode = flushmode;
+      Flags type = Flags.valueOf(flushmode);
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:148]: I think it would be better to pass an instance of the enum to this method, enforcing type safety.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java [hunk 12]
```diff
@@ -48,38 +49,125 @@ public class ContentGenerator {
...
-    this(keySize, bufferSize, bufferSize);
+  /**
+   * Issue Hsync after every write ( Cannot be used with Hflush ).
+   */
+  private final boolean hSync;
+
+  /**
+   * Issue Hflush after every write ( Cannot be used with Hsync ).
+   */
+  private final boolean hFlush;
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:60]: Replacing these with a single instance of `Flags` would simplify the code.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java [hunk 13]
```diff
@@ -48,38 +49,125 @@ public class ContentGenerator {
...
+
+  /**
+   * Type of flags.
+   */
+  public enum Flags {
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:65]: I think `FlushMode` would be a better name.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java [hunk 14]
```diff
@@ -48,38 +49,125 @@ public class ContentGenerator {
...
+    hSync,
+    hFlush,
+    None,
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/ContentGenerator.java:68]: Constants are usually preferred in `CAPS`.

**Review by adoroszlai — COMMENTED** [review]
Thanks @ArafatKhan2198 for updating the PR with the enum from #3192.
